File: LeetCode/Java/239-sliding-window-滑动窗口最大值.java

/*
 * @lc app=leetcode.cn id=239 lang=java
 *
 * [239] 滑动窗口最大值
 */
import java.util.*;

// @lc code=start
class Solution {
    ArrayDeque<Integer> deque = new ArrayDeque();
    int[] nums;

    void cleanDeque(int i, int k) {
        if (!deque.isEmpty() && deque.getFirst() == i - k)
            deque.removeFirst();
        while (!deque.isEmpty() && nums[i] > nums[deque.getLast()])
            deque.removeLast();
    }

    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        if (n * k == 0)
            return new int[0];
        if (k == 1)
            return nums;

        this.nums = nums;
        int maxIdx = 0;
        for (int i = 0; i < k; i++) {
            cleanDeque(i, k);
            deque.addLast(i);
            if (nums[i] > nums[maxIdx])
                maxIdx = i;
        }

        int[] res = new int[n - k + 1];
        res[0] = nums[maxIdx];

        for (int i = k; i < n; i++) {
            cleanDeque(i, k);
            deque.addLast(i);
            res[i - k + 1] = nums[deque.getFirst()];
        }

        return res;
    }
}
// @lc code=end

File: LeetCode/C/239-sliding-window-滑动窗口最大值.cpp
/*
 * @lc app=leetcode.cn id=239 lang=cpp
 *
 * [239] 滑动窗口最大值
 */
#include <bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
  public:
    vector<int> maxSlidingWindow(vector<int> &nums, int k) {
        vector<int> res;
        for (int i = 0; i < k; i++)
            clean_queue(i - k, i, nums);
        res.push_back(nums[dq.front()]);
        for (int i = k; i < nums.size(); i++) {
            clean_queue(i - k, i, nums);
            res.push_back(nums[dq.front()]);
        }
        return res;
    }

  private:
    deque<int> dq;
    void clean_queue(int idx, int i, vector<int> &nums) {
        if (!dq.empty() && idx == dq.front())
            dq.pop_front();
        while (!dq.empty() && nums[dq.back()] < nums[i])
            dq.pop_back();
        dq.push_back(i);
    }
};
// @lc code=end
